### PR TITLE
Cortex-M: move SDPA score scale after bmm

### PR DESCRIPTION
**Pull request by amacharla15**
### Summary

Move the split score scaling produced by decomposed default SDPA from the two BMM inputs to a single scale after the score BMM during Cortex-M `transform_for_annotation`.

The SDPA decomposition produces approximately:

```text
bmm(q * c, k * c)    where c = head_dim^(-1/4)
```

This pass canonicalizes that to:

```text
bmm(q, k) * c^2
```

for the narrow float32 default-SDPA pattern.

This keeps the Cortex-M quantizer unchanged while allowing the stock quantizer to produce proportional qparams across the attention scale. It follows the stage separation discussed in #22796: graph normalization in `transform_for_annotation`, with any later fusion left to the Cortex-M lowering pipeline.

The transformation is a floating-point reassociation and is not bit-equivalent in general, so matching is intentionally restricted to the default decomposed-SDPA score pattern.

### Test plan

* Added pass tests covering default SDPA at multiple head dimensions and conservative rejection cases.
* Added a stock `CortexMQuantizer` regression checking the resulting scale/zero-point relationship.
* Tested head dimensions 16, 32, 64, and 128 across 10 calibration seeds each; all 40 configurations produced compatible qparams, with worst relative scale error `1.20e-7` against a `1e-6` tolerance.
* `lintrunner -a`
* `10 passed` in `test_move_sdpa_scale_after_bmm_pass.py`

Addresses the scalar-scale/qparam portion of #21943.

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218 @mansnils @Sebastian-Larsson @robell @rascani @kimishpatel @jerryzh168 @metascroy

### CODE REVIEW

**Comment by nil-is-all** [comment]
Hi @AdrianLundell, CI looks good on this PR. Could you give a final review?

**Review by AdrianLundell — CHANGES_REQUESTED** [review]
Thanks for the update, this is getting closer now! The new pass is targeting the decomposition from the Arm DecomposeSDPAWithRegularSoftmaxPass and then rewrites it, I see two things wrong with this:
1. As currently written the pass will never do anything since DecomposeSDPAWithRegularSoftmaxPass is not in the Cortex-M pass pipeline.
2. Rewriting the decomposition of one pass with another one is a bit convoluted, it would make more sense to me to reimplement the SDPA decomposition in a way which suits the Cortex-M backend better directly instead.

Please also add a full SDPA operator test with these changes to make sure the full flow makes sense, if there are more changes required to make them pass fully that is fine, just make sure that the intended change is happening (muls are annotated correctly for folding) and then xfail the tests with a comment about remaining work needed.

**Comment by amacharla15** [comment]
Thanks for the review @AdrianLundell. I updated the PR to match the points you raised.

- You pointed out that the previous pass depended on `DecomposeSDPAWithRegularSoftmaxPass`, which is not part of the Cortex-M pipeline, so it never ran in the real flow. I removed that dependency and replaced the old rewrite pass with `DecomposeSDPAPass`, which runs directly in `transform_for_annotation`.

- You also suggested decomposing SDPA directly in a Cortex-M-friendly form instead of rewriting another decomposition. The new pass lowers eligible SDPA directly to:

      matmul -> mul(scale) -> softmax -> matmul

  with the scale applied once to the scores. The existing `MatmulToBmmPass` then converts the supported matmuls to BMMs. Unsupported SDPA forms are left unchanged.

- SDPA operator test: `ops/test_sdpa.py` now starts from `F.scaled_dot_product_attention` and runs through the Cortex-M flow without manually invoking any SDPA decomposition pass.

- You also asked to make sure the intended MUL behavior is present for folding, and to xfail the remaining work if needed. The test verifies that the score MUL sits between DQ and Q with `out_scale ~= in_scale * scale` and equal zero points. The dialect cases that still need the MUL-folding pass are xfailed only on the specific expected failure message, so unrelated failures still fail normally.

One remaining issue: `nn.MultiheadAttention(need_weights=False)` lowers on main today with attention in float. With this change its attention BMMs become quantized, and lowering reaches the separate folded-linear retrace issue from #21943 (`int8` vs `int32`).

File: backends/cortex_m/test/ops/test_sdpa.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+known_dialect_failures: dict[type[torch.nn.Module], tuple[str, str]] = {
+    CortexMSDPA: _SCALE_NOT_FOLDED,
+    CortexMMultiheadAttention: _LINEAR_RETRACED_AS_INT32,
+}
+known_implementation_failures: dict[type[torch.nn.Module], tuple[str, str]] = {
+    CortexMMultiheadAttention: _LINEAR_RETRACED_AS_INT32,
+}
+
+
+def _run_with_known_failure(
```
Comment (AdrianLundell) [inline backends/cortex_m/test/ops/test_sdpa.py]: Please re-use the existing xfail mechanism in the parametrize decorator instead.

**Review by AdrianLundell — COMMENTED** [review]
Looks good overall to me now, just fix the tests to use the regular xfailing mechanism instead of a special function and it is an approve from my side. @rascani Do you have any input also?

**Comment by amacharla15** [comment]
Thanks @AdrianLundell , switched to the xfails mechanism in parametrize.

**Comment by rascani** [comment]
This LGTM, I think I'd just ask that we follow-up with a fix for the `nn.MultiheadAttention(need_weights=False)` regression soon. @amacharla15 - let us know if you're planning on tackling that.

Thanks for this! I really appreciate your contribution.

**Review by rascani — APPROVED** [review]

**Comment by amacharla15** [comment]
Thanks @rascani, really appreciate the review! Yes, I’d like to take the nn.MultiheadAttention(need_weights=False) follow-up.

@AdrianLundell, on #22796 you mentioned an upcoming prototype for changing how folded Q/DQ parameters are handled in the graph. Is that work available now, and should I base this follow-up on it rather than adding a separate Cortex-M-specific workaround?
